rllib/core/rl_module/apis/value_function_api.py

```diff
@@ -0,0 +1,23 @@
+import abc
+from typing import Any, Dict
+
+from ray.rllib.utils.typing import TensorType
+
+
+class ValueFunctionAPI(abc.ABC):
```

Comment:
I get a better idea of how this should look like. I think these APIs (or better abstract interfaces) give users a great structure in how to think of their `RLModules`. This said, it does not simplify the major complexity of `RLModule`s imo as the latter is specifically driven, by the complex setup with a `__init__`, a `__post_init__`, a `setup` and a given call order through multiple parents. 

What becomes more easy is to define modules usable with different `RLlib` algorithms, e.g. `PPO`, `DQN`, etc. We need, however, to give it some thoughts how to best load a state into modules when some parts (e.g. Q function) are not part of the loaded state, but the loading `RLModule`. Here we either need to take care of this in the APIs, or via the `Checkpointable` (checking also for which components in a module are available and in the state, matching these and leaving the rest untouched (initialized by the frameworks initializers). 

I also have a suggestion: if we use these new abstract APIs, it would be nice, if the `inference-only` version is then managed by these APIs themselves. The basic `RLModule`s should have only an actor (or in case of DQN an encoder).

Reply:
Very good points! Let me iterate a bit into a future final design, trying to bring in your thoughts:
* I agree. These new APIs should help (a bit) to ease the burden on the user to think too much up front about the algo being used. For example, in an ideal world, the user would only be concerned about the policy net, producing the actions.
* I agree also that the call orders and having to think about `__init__`, `__post_init__`, and `setup` are still too complex at this point. However, if we could settle on a design, in which the user would only have to override `setup`, then that would be fine.
* The new APIs will help with the complex call order, b/c we can now get rid of multi-inheritance. This is only given the APIs are purely abstract (and thus don't have a constructor themselves <- I think this is crucial and we need to stick to this rule 100%).
* inference-only: This should probably be a) an integral part of RLModule or b) another API, but I tend to prefer a).
* Checkpointing: I think an example would help here: Pre-train a simple (custom) policy-only `RLModule` with BC, then checkpoint a the decently trained policy. Write a sub-class of the policy-only `RLModule`, in which we add a value function net (separate or shared, doesn't matter). If in this subclass we don't touch the original policy-net structure, we should be able to load the checkpoint into this new `ValueFunctionAPI` implementing RLModule. The loaded module will then have the trained (BC) policy and some randomly initialized value function that can be fine-tuned with online e.g. PPO.